client/model.go

```diff
@@ -50,6 +55,114 @@ func (client *Client) Remodel(b []byte) (changeID string, err error) {
 	return client.doAsync("POST", "/v2/model", nil, headers, bytes.NewReader(data))
 }
 
+// RemodelOffline tries to remodel the system with the given model assertion
+// and local snaps and assertion files.
+func (client *Client) RemodelOffline(
+	model []byte, snapPaths, assertPaths []string) (changeID string, err error) {
+
+	// Check if all files exist before starting the go routine
+	snapFiles, err := checkAndOpenFiles(snapPaths)
+	if err != nil {
+		return "", err
+	}
+	assertsFiles, err := checkAndOpenFiles(assertPaths)
+	if err != nil {
+		return "", err
+	}
+
+	pr, pw := io.Pipe()
+	mw := multipart.NewWriter(pw)
+	go sendRemodelFiles(model, snapPaths, snapFiles, assertsFiles, pw, mw)
+
+	headers := map[string]string{
+		"Content-Type": mw.FormDataContentType(),
+	}
+
+	_, changeID, err = client.doAsyncFull("POST", "/v2/model", nil, headers, pr, doNoTimeoutAndRetry)
+	return changeID, err
+}
+
+func checkAndOpenFiles(paths []string) ([]*os.File, error) {
+	var files []*os.File
+	for _, path := range paths {
+		f, err := os.Open(path)
+		if err != nil {
+			for _, openFile := range files {
+				openFile.Close()
+			}
+			return nil, fmt.Errorf("cannot open %q: %w", path, err)
+		}
+
+		files = append(files, f)
+	}
+
+	return files, nil
+}
+
+func createAssertionPart(name string, mw *multipart.Writer) (io.Writer, error) {
+	h := make(textproto.MIMEHeader)
+	h.Set("Content-Disposition",
+		fmt.Sprintf(`form-data; name="%s"`, name))
+	h.Set("Content-Type", asserts.MediaType)
+	return mw.CreatePart(h)
+}
+
+func sendRemodelFiles(model []byte, paths []string, files, assertFiles []*os.File, pw *io.PipeWriter, mw *multipart.Writer) {
```

Comment:
In case of errors in this function, `mw` is not closed, only `pw`. Is this intentional?

Reply:
Yes it is, the reason is that `mw.Close()` is about writing a last boundary and it is not really closing anything (maybe it is not the best name). So it is "closed" only in the success case, otherwise we return immediately an error. This is the same that is done in `client.sendSnapFiles`.

Reply:
Okay thanks for explaining this, it just looks like an oversight when looking at the code without that insight. I think `Closed` is a bit of an unlucky naming for that, as it implies something different. But that is irrelevant to this PR.